### PR TITLE
Add initial caas provisioner worker

### PR DESCRIPTION
**Pull request by wallyworld**
## Description of change

Introduce the CAAS provioner worker and associated facades. Based on initial work by @mjs and @mikemccracken 

For now, any deployed charm will result in no-op operator running in a pod named after the application.

As is the case when adding new facades/workers, there's a lot of boilerplate. The key changes include:

- worker which (for now) deploys an operator running a stock Ubuntu image
- facade which allows to worker to watch  for new applications and ask for config
- restricted CAAS facade support for logins to CAAS models
- new manifold config for CAAS models
- CAAS model entity (includes moving CAAS model tests off ModelSuite)
- driveby fixes to stop errors on CAAS model destroy

## QA steps

juju deploy kubernetes-core
juju add-caas kubernetes caascloud
juju add-model mycaas caascloud
juju deploy <any charm>
kubectl get all

see that a new juju-operator-<appname> is created

### CODE REVIEW

**Comment by howbazaar** [comment]
Please don't land this until we branch 2.3 off.

File: state/caasmodel_test.go [hunk 1]
```diff
@@ -0,0 +1,213 @@
...
+
+	model, modelTag, owner := s.newCAASModel(c)
+	c.Assert(model.Type(), gc.Equals, state.ModelTypeCAAS)
+	c.Assert(model.UUID(), gc.Equals, modelTag.Id())
+	c.Assert(model.Tag(), gc.Equals, modelTag)
+	c.Assert(model.ControllerTag(), gc.Equals, s.State.ControllerTag())
+	c.Assert(model.Owner(), gc.Equals, owner)
+	c.Assert(model.Name(), gc.Equals, "testing")
+	c.Assert(model.Life(), gc.Equals, state.Alive)
+	c.Assert(model.CloudRegion(), gc.Equals, "")
```
Comment (axw) [inline state/caasmodel_test.go]: Seems wrong that we're exposing CloudRegion for CAAS models if it never applies. Can we move that to IAASModel?

File: state/caasmodel_test.go [hunk 2]
```diff
@@ -0,0 +1,213 @@
...
+	err = model.Refresh()
+	c.Assert(err, jc.ErrorIsNil)
+	// TODO(caas) - this will be dying when we add cleanup steps.
+	c.Assert(model.Life(), gc.Equals, state.Dead)
+}
+
+func (s *CAASModelSuite) TestCAASModelsCantHaveCloudRegion(c *gc.C) {
+	s.SetFeatureFlags(feature.CAAS)
+	cfg, _ := s.createTestModelConfig(c)
+	_, _, err := s.State.NewModel(state.ModelArgs{
```
Comment (axw) [inline state/caasmodel_test.go]: Can we please change NewModel to two new methods, NewIAASModel and NewCAASModel, and leave CloudRegion out of the latter? Rename the existing NewModel to NewIAASModel and drop Type from the ModelArgs (also renamed to IAASModelArgs). Then add a new method/type.

File: state/model.go [hunk 3]
```diff
@@ -228,20 +228,21 @@ func (st *State) ModelExists(uuid string) (bool, error) {
 	return count > 0, nil
 }
 
-// ModelActive returns true if a model with the supplied UUID exists
-// and is not being imported as part of a migration.
-func (st *State) ModelActive(uuid string) (bool, error) {
+// ModelActive returns true and the model type if a model
```
Comment (axw) [inline state/model.go]: Please don't overload methods. The model type has nothing to do with whether it's active or not. Create a new method.

File: state/state.go [hunk 4]
```diff
@@ -1060,139 +1060,19 @@ func (st *State) AddApplication(args AddApplicationArgs) (_ *Application, err er
 	if args.Storage == nil {
 		args.Storage = make(map[string]StorageConstraints)
 	}
-	im, err := st.IAASModel()
```
Comment (axw) [inline state/state.go]: starting to get a bit hairy to keep everything paged in at once, can you please split these core State changes into its own PR?

File: worker/caasprovisioner/k8s.go [hunk 5]
```diff
@@ -0,0 +1,175 @@
+// Copyright 2017 Canonical Ltd.
+// Licensed under the AGPLv3, see LICENCE file for details.
+
+package caasprovisioner
```
Comment (axw) [inline worker/caasprovisioner/k8s.go]: I stopped here. While this works, and gets us closer to an MVP, I'm not comfortable pushing forward with this unless we split off into another prototype branch. Please separate the concerns of managing a connection to k8s from provisioning. Perhaps in two separate PRs, so it's clear to which worker/job the parts contribute.

File: state/caasmodel_test.go (hunk 1 above)
Comment (wallyworld) [reply state/caasmodel_test.go]: CAAS models might get regions in the future AFAIUI

Comment (axw) [reply state/caasmodel_test.go]: then we can expose it when it actually means something

File: state/model.go (hunk 3 above)
Comment (wallyworld) [reply state/model.go]: Ok, but we need to know those 2 bits of info at the same time. So we're adding an extra mongo read which introduces load for little benefit. I could rename the method

Comment (axw) [reply state/model.go]: In the *current* code base, we have one place where we call ModelActive, and will also need to know the model type. Lumping together unrelated information in this method is going to cause more harm than an extra mongo read. Calling it "ModelActiveAndModelType" is not really any better.

An alternative would be to pass the StatePool into this worker, get a *state.Model, and use its methods to get the bits that are all read in at once.

File: state/caasmodel_test.go (hunk 1 above)
Comment (wallyworld) [reply state/caasmodel_test.go]: Many, many places use just take a model entity/interface and call model.Region()
We'd have to check for model type all over.
Federated kubernetes supports regions and it's only a matter of time till that will be woven into our caas cloud model

Comment (axw) [reply state/caasmodel_test.go]: OK, then. When dealing with multi-region federated Kubernetes, do you specify a region? Or do the clients just auto connect to the closest API server? Couldn't tell if regions in this case are really something that we should be exposing to users, like we do with IaaS models.
